Python/Product/Analysis/LanguageServer/ILanguageServerExtension.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System.Collections.Generic;
+using Microsoft.PythonTools.Projects;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer {
+    public interface ILanguageServerExtension {
+        /// <summary>
+        /// Called when the extension is registered for a language server.
+        /// </summary>
+        void Register(Server server);
+
+        /// <summary>
+        /// The name of the extension. Used to look up the current instance
+        /// when processing extension command messages. If null or empty,
+        /// the extension cannot be sent messages and may be garbage collected
+        /// if it does not hook any events in its <see cref="Register(Server)"/>
+        /// method.
+        /// </summary>
+        string Name { get; }
+
+        /// <summary>
+        /// Called when an extension command arrives for this extension.
+        /// </summary>
+        Dictionary<string, object> ExecuteCommand(string command, Dictionary<string, object> properties);
```

Comment:
`IReadOnlyDictionary`?

Reply:
Maybe... is Newtonsoft going to serialize it properly? Or will I need to copy it into a real dictionary?

Reply:
Can't we explicitly accept and return `JToken`?

Reply:
OK, if it is simpler for JSON serialization.

Reply:
It looks like Newtonsoft has support for `IReadOnlyDictionary` so I'll go with that.